pkg/tar/tar.go

```diff
@@ -313,23 +325,31 @@ func (t *stiTar) writeTarHeader(tarWriter *tar.Writer, dir string, path string,
 	return nil
 }
 
-// ExtractTarStream calls ExtractTarStreamWithLogging with a nil logger
+// ExtractTarStream calls ExtractTarStreamFromTarReader with a default reader and nil logger
 func (t *stiTar) ExtractTarStream(dir string, reader io.Reader) error {
-	return t.ExtractTarStreamWithLogging(dir, reader, nil)
+	tarReader := tar.NewReader(reader)
+	return t.ExtractTarStreamFromTarReader(dir, tarReader, nil)
 }
 
-// ExtractTarStreamWithLogging extracts files from a given tar stream.
-// Times out if reading from the stream for any given file
-// exceeds the value of timeout
+// ExtractTarStreamWithLogging calls ExtractTarStreamFromTarReader with a default reader
 func (t *stiTar) ExtractTarStreamWithLogging(dir string, reader io.Reader, logger io.Writer) error {
 	tarReader := tar.NewReader(reader)
-	errorChannel := make(chan error)
-	timeout := t.timeout
-	timeoutTimer := time.NewTimer(timeout)
+	return t.ExtractTarStreamFromTarReader(dir, tarReader, logger)
+}
+
+// ExtractTarStreamFromTarReader extracts files from a given tar stream.
+// Times out if reading from the stream for any given file
+// exceeds the value of timeout
+func (t *stiTar) ExtractTarStreamFromTarReader(dir string, tarReader Reader, logger io.Writer) error {
+	errorChannel := make(chan error, 1)
```

Comment:
if the timer fires before the goroutine finishes, the parent select will read from the timer channel and the parent function will then exit.  Nothing hangs around to listen to the errorchannel.  If the errorchannel is unbuffered, then when at some point the goroutine writes its result to it, it will block forever.  Whereas, if buffered, the goroutine will then exit out and the channel will be GCed.

(I think this is the only example of this issue in s2i).